File: src/printfriendly/layout.py
"""
Layout Composer - Apply editorial design system to content.

Preserves original HTML structure (links, paragraphs) while adding
magazine-style presentation with images wrapping text.
"""

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from bs4 import BeautifulSoup, Tag
from jinja2 import Environment, FileSystemLoader, select_autoescape

from .analyzer import AnalyzedContent, ImagePlacement, PullQuote


@dataclass
class LayoutOptions:
    """Options for layout composition."""

    style: str = "magazine"  # magazine, minimal
    include_images: bool = True
    include_pull_quotes: bool = True
    include_drop_cap: bool = True
    include_header_footer: bool = True


class LayoutComposer:
    """Compose editorial layouts from analyzed content."""

    def __init__(self, templates_dir: Optional[Path] = None):
        """Initialize with optional custom templates directory."""
        if templates_dir is None:
            templates_dir = Path(__file__).parent / "templates"

        self.templates_dir = templates_dir
        self.env = Environment(
            loader=FileSystemLoader(str(templates_dir)),
            autoescape=select_autoescape(["html", "xml"]),
        )

    def compose(
        self,
        content: AnalyzedContent,
        options: Optional[LayoutOptions] = None,
    ) -> str:
        """
        Compose the final HTML layout.

        Args:
            content: Analyzed content to layout
            options: Layout options

        Returns:
            Complete HTML document ready for PDF rendering
        """
        if options is None:
            options = LayoutOptions()

        # Process the article body - preserve original HTML!
        body_html = self._compose_body(content, options)

        # Get the main template
        template = self.env.get_template("article.html")

        # Get the styles directory path for font embedding
        styles_dir = Path(__file__).parent / "styles"
        fonts_dir = Path(__file__).parent / "fonts"

        # Render the complete document
        return template.render(
            title=content.title,
            subtitle=content.subtitle,
            author=content.author,
            date=content.date,
            kicker=content.kicker,
            source_name=content.source_name,
            source_url=content.source_url,
            body_content=body_html,
            word_count=content.word_count,
            reading_time=content.reading_time_minutes,
            style=options.style,
            include_header_footer=options.include_header_footer,
            styles_dir=str(styles_dir),
            fonts_dir=str(fonts_dir),
        )

    def _compose_body(self, content: AnalyzedContent, options: LayoutOptions) -> str:
        """Compose the article body preserving HTML structure."""
        # Parse the original HTML content
        soup = BeautifulSoup(content.html_content, "lxml")

        # Find the body or main content
        body = soup.find("body")
        if body:
            # Work with body's children
            main_content = body
        else:
            # Wrap content in a div
            main_content = soup

        # Create image placement map (paragraph_index -> list of placements)
        image_map = {}
        hero_image = None
        for placement in content.image_placements:
            if placement.placement_type == "hero":
                hero_image = placement
            else:
                if placement.paragraph_index not in image_map:
                    image_map[placement.paragraph_index] = []
                image_map[placement.paragraph_index].append(placement)

        # Create pull quote map
        quote_map = {pq.paragraph_index: pq for pq in content.pull_quotes}

        # Build the new body
        new_body = BeautifulSoup("<div class='article-body'></div>", "html.parser")
        article_body = new_body.find("div")

        # Add hero image at the very top
        if hero_image and options.include_images:
            article_body.append(self._create_figure(hero_image, new_body))

        # Process all elements, inserting images and pull quotes
        paragraphs = main_content.find_all(["p", "h1", "h2", "h3", "h4", "h5", "h6",
                                            "blockquote", "ul", "ol", "figure", "pre"])
        para_count = 0
        inserted_quotes = set()

        for elem in paragraphs:
            # Skip empty paragraphs
            if elem.name == "p" and not elem.get_text(strip=True):
                continue

            # Skip figures (we're handling images ourselves)
            if elem.name == "figure":
                continue

            # Clone the element to preserve its structure
            new_elem = self._clone_element(elem, new_body)

            # Add drop cap class to first paragraph
            if para_count == 0 and elem.name == "p" and options.include_drop_cap:
                existing_class = new_elem.get("class", [])
                if isinstance(existing_class, str):
                    existing_class = [existing_class]
                new_elem["class"] = existing_class + ["drop-cap"]

            # Add section-heading class to headings
            if elem.name in ["h1", "h2", "h3", "h4", "h5", "h6"]:
                existing_class = new_elem.get("class", [])
                if isinstance(existing_class, str):
                    existing_class = [existing_class]
                new_elem["class"] = existing_class + ["section-heading"]

            # Check if we should insert an image BEFORE this paragraph (for wrapping)
            if options.include_images and para_count in image_map:
                for placement in image_map[para_count]:
                    fig = self._create_figure(placement, new_body)
                    article_body.append(fig)

            # Add the element
            article_body.append(new_elem)

            # Check if we should insert a pull quote after this paragraph
            if (
                options.include_pull_quotes
                and para_count in quote_map
                and para_count not in inserted_quotes
            ):
                quote = quote_map[para_count]
                article_body.append(self._create_pull_quote(quote, new_body))
                inserted_quotes.add(para_count)

            if elem.name == "p":
                para_count += 1

        # Add any remaining images at the end
        remaining_positions = set(image_map.keys()) - set(range(para_count))
        for pos in sorted(remaining_positions):
            for placement in image_map[pos]:
                if options.include_images:
                    article_body.append(self._create_figure(placement, new_body))

        return str(article_body)

    def _clone_element(self, elem: Tag, soup: BeautifulSoup) -> Tag:
        """Deep clone an element, preserving all children and attributes."""
        # Create a copy of the element
        new_elem = soup.new_tag(elem.name)

        # Copy attributes
        for attr, value in elem.attrs.items():
            new_elem[attr] = value

        # Copy children (including text and nested elements)
        for child in elem.children:
            if isinstance(child, str):
                new_elem.append(soup.new_string(child))
            elif hasattr(child, 'name'):
                # Recursively clone child elements
                new_elem.append(self._clone_element(child, soup))

        return new_elem

    def _create_figure(self, placement: ImagePlacement, soup: BeautifulSoup) -> Tag:
        """Create a figure element with appropriate placement class."""
        # Map placement types to CSS classes
        css_class = f"figure {placement.placement_type}"

        figure = soup.new_tag("figure", attrs={"class": css_class})

        img = soup.new_tag(
            "img",
            attrs={
                "src": placement.image.data_uri,
                "alt": placement.image.alt_text or "Article image",
            },
        )
        figure.append(img)

        # Add caption if available
        caption_text = placement.image.caption or placement.image.alt_text
        if caption_text and caption_text.strip():
            figcaption = soup.new_tag("figcaption")
            figcaption.string = caption_text
            figure.append(figcaption)

        return figure

    def _create_pull_quote(self, quote: PullQuote, soup: BeautifulSoup) -> Tag:
        """Create a pull quote element."""
        aside = soup.new_tag("aside", attrs={"class": "pull-quote"})

        blockquote = soup.new_tag("blockquote")
        blockquote.string = quote.text
        aside.append(blockquote)

        return aside


def compose_layout(
    content: AnalyzedContent,
    style: str = "magazine",
    include_images: bool = True,
    include_pull_quotes: bool = True,
) -> str:
    """Convenience function to compose layout."""
    composer = LayoutComposer()
    options = LayoutOptions(
        style=style,
        include_images=include_images,
        include_pull_quotes=include_pull_quotes,
    )
    return composer.compose(content, options)

File: browser/js/layout.js
/**
 * Layout Composer - Build article DOM structure with editorial design
 * Browser-based port of Python layout.py
 */

const Layout = (function() {
    'use strict';

    /**
     * Layout options
     */
    class LayoutOptions {
        constructor(options = {}) {
            this.style = options.style || 'magazine'; // 'magazine' or 'minimal'
            this.includeImages = options.includeImages !== false;
            this.includePullQuotes = options.includePullQuotes !== false;
            this.includeDropCap = options.includeDropCap !== false;
            this.includeHeaderFooter = options.includeHeaderFooter !== false;
        }
    }

    /**
     * Create the article header element
     */
    function createHeader(content, options) {
        const header = document.createElement('header');
        header.className = 'article-header';

        // Kicker (Category/Topic)
        if (content.kicker) {
            const kicker = document.createElement('span');
            kicker.className = 'kicker';
            kicker.textContent = content.kicker;
            header.appendChild(kicker);
        }

        // Main Headline
        const headline = document.createElement('h1');
        headline.className = 'headline';
        headline.textContent = content.title;
        header.appendChild(headline);

        // Deck / Standfirst
        if (content.subtitle) {
            const deck = document.createElement('p');
            deck.className = 'deck';
            deck.textContent = content.subtitle;
            header.appendChild(deck);
        }

        // Byline
        const byline = document.createElement('div');
        byline.className = 'byline';

        const parts = [];

        if (content.author) {
            const authorSpan = document.createElement('span');
            authorSpan.className = 'author';
            authorSpan.textContent = `By ${content.author}`;
            byline.appendChild(authorSpan);
            parts.push('author');
        }

        if (content.date) {
            if (parts.length > 0) {
                const sep = document.createElement('span');
                sep.className = 'separator';
                sep.textContent = '•';
                byline.appendChild(sep);
            }
            const dateSpan = document.createElement('span');
            dateSpan.className = 'date';
            dateSpan.textContent = content.date;
            byline.appendChild(dateSpan);
            parts.push('date');
        }

        if (content.readingTimeMinutes) {
            if (parts.length > 0) {
                const sep = document.createElement('span');
                sep.className = 'separator';
                sep.textContent = '•';
                byline.appendChild(sep);
            }
            const readingTime = document.createElement('span');
            readingTime.className = 'reading-time';
            readingTime.textContent = `${content.readingTimeMinutes} min read`;
            byline.appendChild(readingTime);
        }

        if (byline.children.length > 0) {
            header.appendChild(byline);
        }

        return header;
    }

    /**
     * Create a figure element for an image placement
     */
    function createFigure(placement) {
        // Handle paired images
        if (placement.placementType === 'paired' && placement.pairWith) {
            const container = document.createElement('div');
            container.className = 'figure-pair';

            // First image
            const fig1 = document.createElement('figure');
            fig1.className = 'figure paired';
            const img1 = document.createElement('img');
            img1.src = placement.image.dataUri;
            img1.alt = placement.image.altText || 'Article image';
            fig1.appendChild(img1);
            container.appendChild(fig1);

            // Second image
            const fig2 = document.createElement('figure');
            fig2.className = 'figure paired';
            const img2 = document.createElement('img');
            img2.src = placement.pairWith.image.dataUri;
            img2.alt = placement.pairWith.image.altText || 'Article image';
            fig2.appendChild(img2);
            container.appendChild(fig2);

            return container;
        }

        // Single figure (hero or centered)
        const figure = document.createElement('figure');
        figure.className = `figure ${placement.placementType}`;

        const img = document.createElement('img');
        img.src = placement.image.dataUri;
        img.alt = placement.image.altText || 'Article image';
        figure.appendChild(img);

        // Add caption if available
        const captionText = placement.image.caption || placement.image.altText;
        if (captionText && captionText.trim()) {
            const figcaption = document.createElement('figcaption');
            figcaption.textContent = captionText;
            figure.appendChild(figcaption);
        }

        return figure;
    }

    /**
     * Create a pull quote element
     */
    function createPullQuote(quote) {
        const aside = document.createElement('aside');
        aside.className = 'pull-quote';

        const blockquote = document.createElement('blockquote');
        blockquote.textContent = quote.text;
        aside.appendChild(blockquote);

        return aside;
    }

    /**
     * Clone an element and all its children
     */
    function cloneElement(elem) {
        return elem.cloneNode(true);
    }

    /**
     * Compose the article body from analyzed content
     */
    function composeBody(content, options) {
        // Parse the HTML content
        const parser = new DOMParser();
        const doc = parser.parseFromString(content.htmlContent, 'text/html');

        // Find body content
        const mainContent = doc.body || doc;

        // Build image placement map (paragraph_index -> list of placements)
        const imageMap = new Map();
        let heroImage = null;

        for (const placement of content.imagePlacements) {
            if (placement.placementType === 'hero') {
                heroImage = placement;
            } else {
                if (!imageMap.has(placement.paragraphIndex)) {
                    imageMap.set(placement.paragraphIndex, []);
                }
                imageMap.get(placement.paragraphIndex).push(placement);
            }
        }

        // Build pull quote map
        const quoteMap = new Map();
        for (const pq of content.pullQuotes) {
            quoteMap.set(pq.paragraphIndex, pq);
        }

        // Create new article body container
        const articleBody = document.createElement('div');
        articleBody.className = 'article-body';

        // Add hero image at the very top
        if (heroImage && options.includeImages) {
            articleBody.appendChild(createFigure(heroImage));
        }

        // Process all elements, inserting images and pull quotes
        const elements = mainContent.querySelectorAll(
            'p, h1, h2, h3, h4, h5, h6, blockquote, ul, ol, figure, pre'
        );

        let paraCount = 0;
        const insertedQuotes = new Set();

        for (const elem of elements) {
            // Skip empty paragraphs
            if (elem.tagName === 'P' && !elem.textContent?.trim()) {
                continue;
            }

            // Skip figures (we're handling images ourselves)
            if (elem.tagName === 'FIGURE') {
                continue;
            }

            // Clone the element to preserve its structure
            const newElem = cloneElement(elem);

            // Add drop cap class to first paragraph
            if (paraCount === 0 && elem.tagName === 'P' && options.includeDropCap) {
                newElem.classList.add('drop-cap');
            }

            // Add section-heading class to headings
            if (/^H[1-6]$/.test(elem.tagName)) {
                newElem.classList.add('section-heading');
            }

            // Check if we should insert an image BEFORE this paragraph
            if (options.includeImages && imageMap.has(paraCount)) {
                for (const placement of imageMap.get(paraCount)) {
                    articleBody.appendChild(createFigure(placement));
                }
            }

            // Add the element
            articleBody.appendChild(newElem);

            // Check if we should insert a pull quote after this paragraph
            if (options.includePullQuotes &&
                quoteMap.has(paraCount) &&
                !insertedQuotes.has(paraCount)) {

                const quote = quoteMap.get(paraCount);
                articleBody.appendChild(createPullQuote(quote));
                insertedQuotes.add(paraCount);
            }

            if (elem.tagName === 'P') {
                paraCount++;
            }
        }

        // Add any remaining images at the end
        for (const [pos, placements] of imageMap) {
            if (pos >= paraCount) {
                for (const placement of placements) {
                    if (options.includeImages) {
                        articleBody.appendChild(createFigure(placement));
                    }
                }
            }
        }

        return articleBody;
    }

    /**
     * Create the article footer
     */
    function createFooter(content) {
        const footer = document.createElement('footer');
        footer.className = 'article-footer';

        const p = document.createElement('p');
        p.textContent = 'Originally published at ';

        const link = document.createElement('a');
        link.href = content.sourceUrl;
        link.className = 'source-link';
        link.textContent = content.sourceName;
        p.appendChild(link);

        footer.appendChild(p);

        return footer;
    }

    /**
     * Main compose function - builds complete article element
     */
    function compose(content, options = {}) {
        const opts = new LayoutOptions(options);

        // Create article container
        const article = document.createElement('article');
        article.className = 'article';

        // Apply style class
        if (opts.style === 'minimal') {
            article.classList.add('style-minimal');
        }

        // Build article structure
        const header = createHeader(content, opts);
        const body = composeBody(content, opts);

        article.appendChild(header);
        article.appendChild(body);

        // Add footer if enabled
        if (opts.includeHeaderFooter) {
            const footer = createFooter(content);
            article.appendChild(footer);
        }

        return article;
    }

    // Public API
    return {
        compose,
        LayoutOptions
    };
})();
